File: mirror/tier2-python/showcase/pattern_exhaustive.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Pattern Matching
Module: pattern_exhaustive.py
Purpose: Demonstrate exhaustive pattern matching
"""

from typing import Any


def exhaustive_match(value: Any) -> str:
    """Exhaustive pattern match on all types"""
    if value is None:
        return "nil"
    elif isinstance(value, bool):
        return "boolean"
    elif isinstance(value, int):
        return "integer"
    elif isinstance(value, float):
        return "float"
    elif isinstance(value, str):
        return "string"
    elif isinstance(value, list):
        return "list"
    elif isinstance(value, dict):
        return "dictionary"
    elif isinstance(value, tuple):
        return "tuple"
    elif isinstance(value, set):
        return "set"
    elif callable(value):
        return "function"
    else:
        return "unknown"


def handle_all_types(value):
    """Handle all type cases"""
    match_type = exhaustive_match(value)
    
    if match_type == "integer":
        return value * 2
    elif match_type == "string":
        return value.upper()
    elif match_type == "list":
        return len(value)
    elif match_type == "dictionary":
        return list(value.keys())
    elif match_type == "function":
        return "callable"
    else:
        return None


result = {
    "type_nil": exhaustive_match(None),
    "type_bool": exhaustive_match(True),
    "type_int": exhaustive_match(42),
    "type_float": exhaustive_match(3.14),
    "type_str": exhaustive_match("hello"),
    "type_list": exhaustive_match([1, 2, 3]),
    "type_dict": exhaustive_match({"a": 1}),
    "handle_int": handle_all_types(10),
    "handle_str": handle_all_types("test"),
}

if __name__ == "__main__":
    print(result)

File: mirror/tier2-ruby/showcase/pattern_exhaustive.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Pattern Matching
# Module: pattern_exhaustive.rb
# Purpose: Demonstrate exhaustive pattern matching

def exhaustive_match(value)
  case value
  when nil
    "nil"
  when true, false
    "boolean"
  when Integer
    "integer"
  when Float
    "float"
  when String
    "string"
  when Array
    "list"
  when Hash
    "dictionary"
  when Proc
    "function"
  else
    "unknown"
  end
end

def handle_all_types(value)
  case exhaustive_match(value)
  when "integer"
    value * 2
  when "string"
    value.upcase
  when "list"
    value.size
  when "dictionary"
    value.keys
  when "function"
    "callable"
  else
    nil
  end
end

result = {
  type_nil: exhaustive_match(nil),
  type_bool: exhaustive_match(true),
  type_int: exhaustive_match(42),
  type_float: exhaustive_match(3.14),
  type_str: exhaustive_match("hello"),
  type_list: exhaustive_match([1, 2, 3]),
  type_hash: exhaustive_match({"a" => 1}),
  handle_int: handle_all_types(10),
  handle_str: handle_all_types("test"),
}

puts result.inspect
